src/Rpc/AuthInterceptor.cs

```diff
@@ -0,0 +1,36 @@
+using Grpc.Core;
+using Grpc.Core.Interceptors;
+using NodeGuard.Data.Repositories.Interfaces;
+
+namespace NodeGuard.Rpc;
+
+public class ServerAuthInterceptor : Interceptor
+{
+    private readonly IAPITokenRepository _apiTokenRepository;
+    
+    public ServerAuthInterceptor(IAPITokenRepository apiTokenRepository)
+    {
+        _apiTokenRepository = apiTokenRepository;
+    }
+    
+public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
+    TRequest request, 
+    ServerCallContext context, 
+    UnaryServerMethod<TRequest, TResponse> continuation)
+    {
+        var token = context.RequestHeaders.FirstOrDefault(x => x.Key == "auth-token")?.Value;
+        if (token == null)
+        {
+            throw new RpcException(new Status(StatusCode.Unauthenticated, "No token provided"));
+        }
+
+        var apiToken = await _apiTokenRepository.GetByToken(token);
+        if (apiToken == null)
+        {
+            throw new RpcException(new Status(StatusCode.Unauthenticated, "Invalid token"));
+        }
+
+        return await continuation(request, context);
```

Comment:
I always suggest to return a var than directly returning a function, it's easier for debugging later or you will have to evaluate in the IDE manually.

Reply:
It's true, but it shouldnt be needed to debug this one. Actually in the docs suggested by dotnet documentation is this way since the continuation is a first class object which represent a function and at the is bypassing it... it's not intended to have a result perse